_SPIKE_.md

```diff
@@ -0,0 +1,543 @@
+# Declarative definition of FOREIGN KEY and UNIQUE constraints
+
+- [Current status](#current-status)
+- [Proposal](#proposal)
+  - [Indexes at property level](#indexes-at-property-level)
+  - [Indexes at model level](#indexes-at-model-level)
+  - [Foreign keys at property level](#foreign-keys-at-property-level)
+  - [Foreign keys at model level](#foreign-keys-at-model-level)
+  - [Decorators](#decorators)
+  - [Additional changes](#additional-changes)
+- [Next steps](#next-steps)
+- [Detailed description of the current syntax](#detailed-description-of-the-current-syntax)
+
+## Current status
+
+UNIQUE index is supported as a specialized version of a generic index feature.
+
+Support for indexes and foreign keys depends on the connector. Some do support
+all features, some support only some of them. Depending on the database used,
+certain features are not possible to implement at all because of database design
+constraints. For example, CouchDB/Cloudant do not support UNIQUE indexes
+
+We have a syntax for defining (unique) indexes described in
+[our docs](https://loopback.io/doc/en/lb3/Model-definition-JSON-file.html#indexes),
+this syntax supports composite indexes. However, many connectors are adding
+their own extensions and PostgreSQL uses a slightly different format violating
+our docs.
+
+We have an undocumented syntax for defining foreign keys which is recognized by
+all SQL connectors that provide migration of foreign keys. Unfortunately, this
+syntax does not support composite foreign keys and there is also no support for
+expressing `ON UPDATE` and `ON DELETE` behavior.
+
+You can find more details in
+[Detailed description of the current syntax](#detailed-description-of-the-current-syntax)
+at the bottom.
+
+## Proposal
+
+### Indexes at property level
+
+Support the following two short-hand forms only. Ask users to use model-level
+form to define indexes that are more complex.
+
+- a "plain" index with no special configuration
+
+  ```js
+  {
+    email: {
+      type: 'string',
+      index: true,
+    }
+  }
+  ```
+
+- UNIQUE index with no special configuration
+
+  ```js
```

Comment:
In the case of Cloudant and CouchDB, they don't have UNIQUE index as you mentioned... If I'm using Cloudant as the datasource and happen to have `unique` set. What's going to happen? We'll ignore it because it's not supported by the database?

Reply:
I think the only unique field in cloudant/couchdb is the document id `_id`, other loopback flavor constrains will be ignored.

See https://forums.couchbase.com/t/can-i-do-unique-constraint-on-some-attribute-not-the-id/5621
and https://stackoverflow.com/questions/1541239/unique-constraints-in-couchdb